File: apps/java/SDG/ObserverSwarm.java
import swarm.Globals;
import swarm.objectbase.Swarm;
import swarm.activity.Activity;
import swarm.activity.Schedule;
import swarm.activity.ScheduleImpl;
import swarm.simtoolsgui.GUISwarmImpl;
import swarm.space.Object2dDisplay;
import swarm.space.Object2dDisplayImpl;
import swarm.gui.ZoomRasterImpl;
import swarm.gui.ZoomRaster;
import swarm.gui.Colormap;
import swarm.gui.ColormapImpl;
import swarm.defobj.Zone;
import swarm.Selector;

import java.util.List;
import java.util.LinkedList;

public class ObserverSwarm extends GUISwarmImpl {
  public final static byte UserTourColor = 0;
  public final static byte UserResistColor = 1;
  public final static byte UserListenColor = 2;
  public final static byte GlenTourColor = 3;
  public final static byte GlenAttackColor = 4;
  public final static byte MarcusIncubateColor = 5;
  public final static byte MarcusResistColor = 6;
  public final static byte MarcusListenColor = 7;
  public final static byte MarcusNativeColor = 8;
  public final static byte AlexTourColor = 9;
  public final static byte AlexTalkColor = 10;
  public final static byte AlexTargetColor = 11;
  Object2dDisplay display;
  ZoomRaster raster;
  SDG model;
  Colormap colormap;
  Schedule displaySchedule;

  public ObserverSwarm (Zone aZone) {
    super (aZone);
    model = new SDG (aZone);
  }

  public Object buildObjects () {
    super.buildObjects ();

    model.buildObjects ();
    
    colormap = new ColormapImpl (getZone ());
    colormap.setColor$ToName (UserTourColor, "white");
    colormap.setColor$ToName (UserResistColor, "gray");
    colormap.setColor$ToName (UserListenColor, "slate gray");

    colormap.setColor$ToName (MarcusIncubateColor, "yellow");
    colormap.setColor$ToName (MarcusResistColor, "red");
    colormap.setColor$ToName (MarcusListenColor, "pale green");
    colormap.setColor$ToName (MarcusNativeColor, "dark green");

    colormap.setColor$ToName (GlenTourColor, "cyan");
    colormap.setColor$ToName (GlenAttackColor, "pink");

    colormap.setColor$ToName (AlexTourColor, "brown");
    colormap.setColor$ToName (AlexTalkColor, "orange");
    colormap.setColor$ToName (AlexTargetColor, "brown");

    raster = new ZoomRasterImpl (getZone (), "raster");
    raster.setColormap (colormap);
    raster.setZoomFactor (3);
    raster.setWidth$Height (model.getWorld ().getSizeX (),
                            model.getWorld ().getSizeY ());
    raster.setWindowTitle ("SDG World");
    raster.pack ();
    raster.erase ();

    try {
      display = new Object2dDisplayImpl
        (getZone (),
         raster,
         model.getWorld (),
         new Selector (Class.forName ("agent2d.Agent2d"), "drawSelfOn", false));
      raster.setButton$Client$Message (3, display,
                                       new Selector (display.getClass (),
                                                     "makeProbeAtX$Y", true));
    } catch (Exception e) {
      e.printStackTrace (System.err);
      System.exit (1);
    }


    return this;
  }

  public void updateDisplay () {
    raster.erase ();
    display.display ();
    raster.drawSelf ();
    getActionCache ().doTkEvents ();
    Globals.env.probeDisplayManager.update ();
  }

  public Object buildActions () {
    super.buildActions ();

    model.buildActions ();
    displaySchedule = new ScheduleImpl (getZone (), 1);

    try {
      displaySchedule.at$createActionTo$message
        (0, 
         this,
         new Selector (getClass (), "updateDisplay", false));
    } catch (Exception e) {
      e.printStackTrace (System.err);
      System.exit (1);
    }
    return this;
  }

  public Activity activateIn (Swarm context) {
    super.activateIn (context);
    
    model.activateIn (this);
    displaySchedule.activateIn (this);
    return getActivity ();
  }

  public static void main (String[] args) {
    Globals.env.initSwarm ("SDG", "0.0", "bug-swarm@swarm.org", args);
    ObserverSwarm observer = new ObserverSwarm (Globals.env.globalZone);
    observer.buildObjects ();
    observer.buildActions ();
    observer.activateIn (null);
    observer.go ();
  }
}

File: apps/objc/schellingII/ObserverSwarm.m
// Schellings Segregation Model a la Axtell and Epstein
// Code by Benedikt Stefansson, <benedikt@ucla.edu>. 
// First version July 1997


#import "ObserverSwarm.h"
#import "ModelSwarm.h"
#import <collections.h>
#import <objectbase.h>

@implementation ObserverSwarm

// createBegin: here we set up the default observation parameters.
+createBegin: (id) aZone {
  ObserverSwarm * obj;
  id <ProbeMap> probeMap;
  
  // Superclass createBegin to allocate ourselves.
  obj = [super createBegin: aZone];

  // Fill in the relevant parameters (only one, in this case).
  obj->displayFrequency = 1;

  // Also, build a customized probe map. Without a probe map, the default
  // is to show all variables and messages. Here we choose to
  // customize the appearance of the probe, give a nicer interface.
  probeMap = [EmptyProbeMap createBegin: aZone];
  [probeMap setProbedClass: [self class]];
  probeMap = [probeMap createEnd];

  // Add in a bunch of variables, one per simulation parameters
  [probeMap addProbe: [probeLibrary getProbeForVariable: "displayFrequency"
				    inClass: [self class]]];

  // Now install our custom probeMap into the probeLibrary.
  [probeLibrary setProbeMap: probeMap For: [self class]];

  return obj;
}


// createEnd: create objects we know we'll need. In this case, none,
// but you might want to override this.
-createEnd {
  return [super createEnd];
}

// Create the objects used in the display of the model. 
-buildObjects {
  id modelZone;					  // zone for model.

  [super buildObjects];

  modelZone = [Zone create: self];
  modelSwarm = [ModelSwarm create: modelZone];
  
 
  CREATE_ARCHIVED_PROBE_DISPLAY (modelSwarm);
  CREATE_ARCHIVED_PROBE_DISPLAY (self);


  [controlPanel setStateStopped];

  // Check now if the user hit the quit button: if so, abort.
  if ([controlPanel getState] == ControlStateQuit) return self;

  [modelSwarm buildObjects];


  colormap = [Colormap create: self];

  // set up the colormap, it maps colors used
  // by the Tk widgets - which have actual names -
  // into our internal representation as arbitrary integers
  [colormap setColor: 0 ToName: "blue"];
  [colormap setColor: 1 ToName: "red"];
  [colormap setColor: 2 ToName: "green"];
  [colormap setColor: 3 ToName: "yellow"];
  [colormap setColor: 4 ToName: "purple"];
  [colormap setColor: 5 ToName: "cyan2"];
  [colormap setColor: 6 ToName: "LightSkyBlue3"];
  [colormap setColor: 7 ToName: "SlateGray4"];
  [colormap setColor: 8 ToName: "MidnightBlue"];
  [colormap setColor: 9 ToName: "DarkKhaki"];
  [colormap setColor: 10 ToName: "LightGoldenrod"];
  [colormap setColor: 11 ToName: "IndianRed"];
  [colormap setColor: 12 ToName: "DarkOrange"];
  [colormap setColor: 13 ToName: "HotPink"];
  [colormap setColor: 14 ToName: "SteelBlue2"];
  [colormap setColor: 15 ToName: "VioletRed"];
  [colormap setColor: 16 ToName: "turquoise4"];
  [colormap setColor: 17 ToName: "burlywood2"];
  [colormap setColor: 18 ToName: "gray70"];
  [colormap setColor: 19 ToName: "thistle1"];


  [colormap setColor: 21 ToName: "white"];

  // create the 2d window which is used to
  // display the position of agents on the toroid
  worldRaster = [ZoomRaster createBegin: self];
  SET_WINDOW_GEOMETRY_RECORD_NAME (worldRaster);
  worldRaster = [worldRaster createEnd];
  [worldRaster setColormap: colormap];
  [worldRaster setZoomFactor: 6];
  [worldRaster setWidth: [modelSwarm getWorldSize] Height: [modelSwarm getWorldSize]];
  [worldRaster setWindowTitle: "Schelling's world"];
  [worldRaster pack];

  // Create Object2dDisplay which knows how to
  // display a 2d space of objects/agents and
  // respond to mouseclicks from the raster
  worldDisplay = [Object2dDisplay createBegin: self ];
  [worldDisplay setDisplayWidget: worldRaster];
  [worldDisplay setDiscrete2dToDisplay: [[modelSwarm getWorld] getObjectGrid]];
  [worldDisplay setObjectCollection: [modelSwarm getAgentList]];
  [worldDisplay setDisplayMessage: M(drawSelfOn:)];   
  worldDisplay = [worldDisplay createEnd];

  // This allows the user to right-click on the display to probe agents
  [worldRaster setButton: ButtonRight Client: worldDisplay Message: M(makeProbeAtX:Y:)];

  // Finally add one graph widget to show how many
  // people are unhappy and trying to move
  moveGraph = [EZGraph createBegin: self];
  SET_WINDOW_GEOMETRY_RECORD_NAME (moveGraph);
  [moveGraph setTitle: "Fraction of people unhappy"];
  [moveGraph setAxisLabelsX: "time" Y: "number"];
  moveGraph = [moveGraph createEnd];

  [moveGraph createAverageSequence: "unhappy"
  withFeedFrom: [modelSwarm getAgentList]
  andSelector: M(getUnhappy)] ;


  return self;
}  

-buildActions {
  [super buildActions];

  // Tell the model Swarm to build it's schedules
  [modelSwarm buildActions];

  // Then create a group of actions for the observer
  displayActions = [ActionGroup create: self ];

  [displayActions createActionTo: probeDisplayManager message: M(update)];
  [displayActions createActionTo: actionCache         message: M(doTkEvents)];
  [displayActions createActionTo: self                message: M(eraseRaster)];
  [displayActions createActionTo: worldDisplay        message: M(display)];
  [displayActions createActionTo: worldRaster         message: M(drawSelf)];
  [displayActions createActionTo: moveGraph	message:M(step)];

  // Put these actions on a schedule to be repeated at a certain frequency
  displaySchedule = [Schedule createBegin: self];
  [displaySchedule setRepeatInterval: displayFrequency]; // note frequency!
  displaySchedule = [displaySchedule createEnd];
  [displaySchedule at: 0 createAction: displayActions];
  
  return self;
}  

// activateIn: - activate the schedules so they're ready to run.
// The swarmContext argument has to do with what we were activated *in*.
// Typically the ObserverSwarm is the top-level Swarm, so it's activated

// in "nil". But other Swarms and Schedules and such will be activated
// inside of us.
-activateIn: (id) swarmContext {
  // First, activate ourselves (just pass along the context).
  [super activateIn: swarmContext];

  // Now activate our schedule in ourselves. This arranges for the
  // execution of the schedule we built.
  [displaySchedule activateIn: self];

  // Activate the model swarm in ourselves. The model swarm is a
  // subswarm of the observer swarm.
  [modelSwarm activateIn: self];

  // Activate returns the swarm activity - the thing that's ready to run.
  return [self getSwarmActivity];
}

-eraseRaster {
  // A method that overrides the default
  // black background for raster and paints
  // the raster white instead
  [worldRaster fillRectangleX0: 0
	                    Y0: 0
	                    X1: [modelSwarm getWorldSize] 
			    Y1: [modelSwarm getWorldSize]
	                 Color: 21]; 
	return self;

}


@end
